kanto_kuska/edge_client.py

```diff
@@ -0,0 +1,150 @@
+        self.mqtt_host = mqtt_host
+        self.mqtt_port = mqtt_port
+        self.mqtt_username = mqtt_username
+        self.mqtt_password = mqtt_password
+        self.mqtt_client = mqtt.Client()
+
+        # Set MQTT username and password if provided
+        if mqtt_username and mqtt_password:
+            self.mqtt_client.username_pw_set(mqtt_username, mqtt_password)
+
+        # Subscribe to specified MQTT topics
+        self.subscribe_to_topics()
+
+    def subscribe_to_topics(self):
+        def on_message(client, userdata, message):
+            topic = message.topic
+            payload = json.loads(message.payload)
+
+            # Update internal state based on received topic and payload
+            if topic in self.kuksa_client.values:
+                self.kuksa_client.values[topic] = payload
+                self.handle_kuksa_update(topic, payload)
+
+            print(f"Received message on topic '{topic}': {payload}")
+
+        self.mqtt_client.on_message = on_message
+        self.mqtt_client.connect(self.mqtt_host, self.mqtt_port)
+        self.mqtt_client.loop_start()
+
+        # Subscribe to Kuksa data topics
+        topics_to_subscribe = [
+            "Altitude",
+            "Latitude",
+            "Longitude",
+            "Speed"
+        ]
+        for topic in topics_to_subscribe:
+            self.mqtt_client.subscribe(topic)
+
+    def handle_kuksa_update(self, topic, value):
+        # Handle Kuksa data signal (simplified)
+        print(f"Received Kuksa data signal - Topic: {topic}, Value: {value}")
+        # Trigger Ditto data update
+        self.update_ditto_feature(topic, value)
+
+    def update_ditto_feature(self, property_path, value):
+        # Update Ditto feature with the new value
+        feature_id = "VSS"  # Example feature ID (Vehicle Speed and Location)
+        cmd_envelope = f"{feature_id}:{property_path}:{value}"
+        self.ditto_client.send(cmd_envelope)
+
+    def start(self):
+        try:
+            # Simulate device info retrieval
+            self.device_info = EdgeDeviceInfo()
+            self.device_info.unmarshal_json('{"deviceId": "123", "tenantId": "456", "policyId": "789"}')
+
+            # Add VSS feature to mock Ditto client
+            self.add_vss_feature()
+
+            # Simulate Kuksa data updates
+            self.kuksa_client.simulate_updates()
+
+        except Exception as e:
+            print(f"Error occurred during start: {e}")
+
+    def add_vss_feature(self):
+        # Add VSS feature to mock Ditto client (simplified)
+        feature_id = "VSS"
+        feature_data = {
+            "description": "Vehicle Speed and Location",
+            "properties": self.kuksa_client.values.copy()  # Include a copy of Kuksa values in feature data
+        }
+        self.ditto_client.features[feature_id] = feature_data
+        print(f"Mock Ditto: Added feature '{feature_id}' with properties")
+        print(feature_data)
+
+    def stop(self):

```

Comment:
This method is currently not used. Add detection of terminating signals in the main and use it.